### ISSUE TITLE
Request For Comment: Close sourcing this API and providing an SDK

### ISSUE THREAD

**Issue by jpmcb** [open-sauced/api#799]
As OpenSauced's services and foundation grow, the @open-sauced/engineering team is considering close sourcing this API source code. This is a request for comment (RFC) since we'd like to engage with the community to help us get a better understanding of the impact!

## Why are we considering close sourcing open-sauced/api?

1. The complexity of making open source contributions into this code base from the broader OpenSauced community has grown significantly. Today, in order to bootstrap and get started with running the API, you need an active and populated Timescale database for events, a postgres database for users/metadata, an OpenAI API key, Supabase, a Bing search API key, and more. This is a pretty heavy lift and we haven't seen any active contributions in some time. Further, much of the necessary infrastructure (like the database schema for Timescale) is closed source. This means that it's effectively impossible for an non-OpenSauced employee to run the API without a connection that TypeORM can validate against an accurate schema in Timescale. This decision would more or less be bringing things closer in line with the reality of running the API as an outside contributor.
2. Operationally, this API has reached a point in its maturity where the engineering team needs to consider fewer bespoke "open" solutions and more internal operational solutions for managing our infrastructure and code. Because this API integrates so deeply with our entire product stack, it's a bit awkward trying to manage an open source repository while, behind the scenes, we try to manage a deeply integrated stack that doesn't make sense to bring into the open. For reference,  https://github.com/open-sauced/api/issues/722 highlights some of the difficulty in managing these two different (and often competing) worlds. This means that if we close sourced this codebase, OpenSauced would be able to integrate the API vertically with more services, execute on operational DevOps tasks not appropriate for the open source, lift and shift our internal operations solutions to more "infra-as-code", move faster, and ship more features to client consumers.
3. In order to protect the intellectual "brain-trust" of OpenSauced, we have considered moving the most sensitive bits of future OpenSauced platform architectures and implementations to separate microservices, outside of this API. This has already happened abit with several of our data pipelines and AI microservices. But, for our small engineering team, this more "monolithic" API works really well as a centralized "hub" where many service spokes can be integrated. Splitting this API, with the current team bandwidth, into small microservices in service of protecting the most sensitive intellectual ideas would be ok, but for ease of our operations, we'd likely want to go down the road of close sourcing the entire API codebase so we can continue to move fast here. 

## What does this mean for code consumers of the API?

If you are a "code consumer" of this codebase (i.e., you run the OpenSauced API somewhere, you utilize some of the source code in your project/product as a downstream consumer, you have built additional applications on-top of the OpenSauced source code, etc.) - **we want to hear from you!** Currently, we are unaware of _**any**_ direct code consumers.

If we don't hear from anyone in the coming weeks / months, we will make a decision and proceed with close sourcing this codebase. Any commits, changes, or content will continue to be licensed under the [MIT license](https://github.com/open-sauced/api#MIT-1-ov-file) up to the point of the close sourcing.

## What does this mean for clients of the API?

Nothing! The API service itself will continue to be public and used by [the OpenSauced platform](https://app.opensauced.pizza/). We also plan to continue providing a public swagger doc: https://api.opensauced.pizza/ that can be used to explore the API's controllers and routes.

## Envisioning the OpenSauced SDK

In the future, we plan to build an OpenSauced SDK: this would be a single TypeScript package / library that anyone could use to integrate with the OpenSauced public API. This, like many other SDKs, would be open sourced and would become the defacto place to build ontop of OpenSauced and use the API.

**Comment by bdougie** [comment]
I appreciate you opening this up @jpmcb. We have come a long way from the first commit here and now that we are supporting 10k+ users and growing, it makes sense to approach a different solution for the community while still keeping things open. 

Some observations
- We rarely get contributions to the API.
- The burden of setting up the API locally using multiple services is enough to recommend against it. 
- Having openapi and a future public SDK seems like a great transition. 

Would love to hear more from the community before we make any permanent decisions.

**Comment by takanome-dev** [comment]
I completely agree with @bdougie here. I think this is the right move. I've been discouraged many times trying to set up the server to make contributions, even before we started implementing the Timescale database. I've even received messages from people asking for help with setup because they saw my contributions.

I want to say thank you to all of you for your patience, help, support, and guidance. I have learned a lot from contributing to this project and still come back to it from time to time to learn and get inspired by how things are done. 

Thank you! 🙏🏿

**Comment by brandonroberts** [comment]
I agree with the thoughts shared here. I know there's a stigma of closing source on something that was once open, but I think it would be better for all involved by providing an SDK and open access to the API. 

Our stack has grown significantly over the lifetime of this project, and at this pace the contribution experience could continue to be sub-optimal. I'd rather not be open just for the sake of saying we are without actually having a constructive path for people to contribute.

**Comment by Anush008** [comment]
Hey guys, thanks for laying out the rationale.

So there's the obvious downside of losing out on community contributions. I understand it's hard to make major contributions to the repo, but small tweaks and patches could still make it.

So, my question is, why not continue with your internally governed development flow without flipping the private switch?  Because I don't see any upside in doing this.

How about adding a disclaimer that the repo is open for educational purposes to avoid the overhead of maintaining an interface for contributions? Because in the end, you're not obligated to do so.

This way, the repo can be public for the sake of knowledge sharing at least.

**Comment by jpmcb** [comment]
Thanks for your @Anush008 - always appreciate hearing from you!

> So, my question is, why not continue with your internally governed development flow without flipping the private switch? Because I don't see any upside in doing this.

If we want to vertically integrate more of the API's stack (like adding infrastructure as code tooling, secrets management, cloud provider tools, etc.) and keep this repo public, for the sake of providence, security and sustainability, the open-sauced engineering team would need to take on alot of operational burden in order to piecemeal something together. For example: let's say we wanted to setup our secrets manager to automatically populated a dev, beta, or production `.env` for the engineering team to work against. This is something we already need since the `.env` has gotten pretty unwieldy these days.

Today, with the repo being public, we'd have to come up with a way to do this in some other separate repo where we'd have scripts that populate a `.env` and pull down secrets from our secrets manager at the given vault/path. It would be a mistake to add this sort of script to a public repo since it'd be revealing way too much about the internal landscape of how our secrets manager works / is setup.

This script would then need to know where the local developer's API directory is and populate the `.env` there. There's also a nasty problem of code drift between the 2 repos: what happens if an expected secret changes in the API logic but we missed updating the "operations-repo" and created some drift? You could fix for this with git submodules, more intense dependency management, etc. All of this would be alot of code to manage just to get a script that could populate some secrets. We'd also probably want to put some additional mechanisms in place to ensure that the API code wasn't leaking/committing those secrets.

But if the repo was private, we could just have 1 or 2 small scripts that do this and access the secrets manager with the given vault/path to the secrets directly in the code repo without having the headache of drifting scripts, managing public visibility, etc.

Further, operationally, there is already a great security risk our small team is taking on by open sourcing the most sensitive bits of our API: I.e the controller's logic and database logic. Today, if a threat actor wanted to understand the landscape of how we interface with our data, our user's sensitive information, and more, all they have to do is read the API code. While we already follow best practices, by close sourcing, we gain abit of obfuscation that would make this type of discover much more difficult for threat actors.

---

> This way, the repo can be public for the sake of knowledge sharing at least.

Keeping things open source for educational purposes would be an anti-goal of this part of the proposal:

> > In order to protect the intellectual "brain-trust" of OpenSauced, we have considered moving the most sensitive bits of future OpenSauced platform architectures and implementations to separate micro-services, outside of this API.

One thing we had also considered is moving this to a BSL license (and granted, I'm not a lawyer) but that wouldn't prevent others from reading our codebase and deriving _how_ we do some of the things that should _probably_ remain as internal, intellectual property. I don't think there's much in the API _currently_ that fits that category, but in the future, I personally believe more and more of the API will become the heart and soul of what makes OpenSauced tick and where the "secret sauce" lives.

**Comment by Anush008** [comment]
Wow. Thanks for that practical justification John. I'm convinced 🫡.
